lib/rubocop/cop/rspec/named_subject.rb

```diff
@@ -62,14 +99,53 @@ def on_block(node) # rubocop:disable InternalAffairs/NumblockHandler
           end
 
           subject_usage(node) do |subject_node|
-            add_offense(subject_node.loc.selector)
+            check_explicit_subject(subject_node)
           end
         end
 
+        private
+
         def ignored_shared_example?(node)
           cop_config['IgnoreSharedExamples'] &&
             node.each_ancestor(:block).any?(&method(:shared_example?))
         end
+
+        def check_explicit_subject(node)
+          return if allow_explicit_subject?(node)
+
+          add_offense(node.loc.selector)
+        end
+
+        def allow_explicit_subject?(node)
+          !always? && !named_only?(node)
+        end
+
+        def always?
+          style == :always
+        end
+
+        def named_only?(node)
+          style == :named_only &&
+            subject_definition_is_named?(node)
+        end
+
+        def subject_definition_is_named?(node)
+          subject = nearest_subject(node)
```

Comment:
I wonder about the performance of this two-way ast traversal on each `subject` reference.

Reply:
We do this in a couple of other cops, like `EmptyExampleGroup`, `ImplicitBlockExpectation` just to name a few.

Reply:
Ping

Reply:
I just tried running [RuboCop's profiler](https://github.com/rubocop/rubocop/blob/master/bin/rubocop-profile) on the specs a large repository I work on. And the performance of these AST traversals seem to be fine.